engine/consensus/matching/engine.go

```diff
@@ -307,31 +356,53 @@ func (e *Engine) onReceipt(originID flow.Identifier, receipt *flow.ExecutionRece
 	//  => drop Receipt
 	sealed, err := e.state.Sealed().Head()
 	if err != nil {
-		return fmt.Errorf("could not find sealed block: %w", err)
+		return false, fmt.Errorf("could not find sealed block: %w", err)
 	}
 
 	isSealed := head.Height <= sealed.Height
 	if isSealed {
 		log.Debug().Msg("discarding receipt for already sealed and finalized block height")
-		return nil
+		return false, nil
 	}
 
-	// if the receipt is not valid, because the parent result is unknown, we will drop this receipt.
-	// in the case where a child receipt is dropped because it is received before its parent receipt, we will
-	// eventually request the parent receipt with the `requestPending` function
-	childSpan := e.tracer.StartSpanFromParent(receiptSpan, trace.CONMatchOnReceiptVal)
-	err = e.receiptValidator.Validate([]*flow.ExecutionReceipt{receipt})
-	childSpan.Finish()
+	// we don't want to validate the same receipt more than once,
+	// if this receipt has been stored in receipts database, it must have been
+	// validated, either by ourselve or the consensus committee, so we could skip
+	// validating it.
+	_, err = e.receiptsDB.ByID(receiptID)
 	if err != nil {
-		return fmt.Errorf("failed to validate execution receipt: %w", err)
+		// it's a known receipt, skip processing
+		childSpan := e.tracer.StartSpanFromParent(receiptSpan, trace.CONMatchOnReceiptVal)
+		err = e.receiptValidator.Validate([]*flow.ExecutionReceipt{receipt})
+		childSpan.Finish()
+
+		if validation.IsMissingPreviousResultError(err) {
```

Comment:
did you consider using `UnverifiableInputError`?

Reply:
Yeah, we could. I could refactor that in 387